sequencer/src/transaction.rs

```diff
@@ -6,6 +6,26 @@ use hotshot_types::traits::block_contents::Transaction as HotShotTransaction;
 use jf_merkle_tree::namespaced_merkle_tree::{Namespace, Namespaced};
 use serde::{Deserialize, Serialize};
 
+/// TODO [`NamespaceId`] has historical debt to repay:
+/// - It must fit into 4 bytes in order to maintain serialization compatibility
+///   for [`crate::block::NsTable`], yet it currently occupies a `u64`. Thus,
+///   any code that creates a [`NamespaceId`] must ensure that it fits into 4
+///   bytes, which is bug-prone. We should either (i) allow it to occupy 8
+///   bytes, breaking serialization compatibility, or (ii) define it as a 4-byte
+///   struct such as `u32`.
```

Comment:
So it seems like we're going with the 4-byte restriction for now. We can lift this and switch to 8 bytes as soon as we have a solution for upgrading data structures.

In the meantime, though, why not just change this to `u32` right now? This will prevent bugs and errors. As is, I think we at least have to update the deserialization to fail if the input doesn't fit in a `u32`, or add back the check in the submit endpoint to enforce that nobody is submitting transactions with namespaces that don't fit. Otherwise we will have transactions in the mempool that cause payload construction to fail, and the builder is not smart enough to filter those out.

Reply:
> why not just change this to `u32` right now?

I strongly prefer this solution. I didn't put it in this PR because I wasn't sure whether it would introduce a serialization break. (And because this PR is already way too big.) I don't even know whether `reference_tests.rs` would catch a change in `Transaction` serialization.

But now that I have this feedback I'll give it a shot.

Reply:
Ah, shoot I think you're right actually. When we serialize a transaction by itself, outside of a payload, this might be a break. For JSON it definitely won't matter. For bincode we might get away with it if using a variable length integer encoding. I'm not sure if that's turned on by default, so it's worth trying.

If we can't make the `u32` switch work, I think we should override *de*serialization so it parses a u64 but returns an error if it exceeds `u32::MAX`.

Reply:
Confirmed: changing it to `u32` breaks reference tests. 😕 

> If we can't make the `u32` switch work, I think we should override _de_serialization so it parses a u64 but returns an error if it exceeds `u32::MAX`.

Could do. But that won't stop anyone from using `NamespaceId::from` on an arbitrary `u64`, which is what we do in `endpoints.rs`. How would you feel about removing `From` from the `derive` list for `NamespaceId`? (There's a bunch of other cruft there I'd like to remove, too. See the rest of this code comment block.)

Reply:
Ah, I didn't realize we were using `NamespaceId::from`, I thought we were just deserializing the whole transaction in the endpoint.

Anyway, yeah, how about we replace `From<u64>` with `From<u32>` and that will place the typechecking/range checking burden on whoever is trying to create a namespace ID?

Reply:
Oh, we have two endpoints that deal with namespace IDs: the `submit` endpoint just deserializes a whole transaction (which recursively calls deserialize on the namespace ID). The namespace proof endpoint uses `NamespaceId::from` directly. So I think if we fix both deserialize and `From`, we will be safe

Reply:
`serde` deserialize fixed in https://github.com/EspressoSystems/espresso-sequencer/pull/1499/commits/fed11618f98fdca3a1f738fc4ab5de74875db0d3

`From<u32>` fixed in https://github.com/EspressoSystems/espresso-sequencer/pull/1499/commits/26a6b299ce260a210e6fa3253bbd50fb57f3023d

I prefer to drop `From<u32>` and force the caller to rely solely on `serde`. But at first glance it seems that `serde` is invoked only by `body_auto` in [RequestParams in tide_disco::request - Rust](https://tide-disco.docs.espressosys.com/tide_disco/request/struct.RequestParams.html#method.body_auto), which seems to want to parse the entire request body. By contrast, if you want only an individual param then you need to use `param`, which doesn't seem to use `serde` (?) and instead converts directly to fundamental types such as `u64`, which you must then deserialize into a `NamespaceId`. Instead of fighting with it I just implemented both solutions as you suggest.

Wow this is ugly. New issue https://github.com/EspressoSystems/espresso-sequencer/issues/1574